### PR TITLE
feat(dev): `ignoredRouteFiles` config option

### PR DESCRIPTION
**Pull request by chaance**
This PR introduces a new configuration API for ignoring non-route files in `app/routes`. This would allow users to keep CSS or test files adjacent to their route modules. 

```js
// remix.config.js
module.exports = {
  // ...
  // new optional config option
  ignoredRouteFiles: ["*.css"]
};
```

### CODE REVIEW

**Review by mcansh — APPROVED** [review]

File: packages/remix-dev/config.ts [hunk 1]
```diff
@@ -83,6 +83,12 @@ export interface AppConfig {
    * The output format of the server build. Defaults to "cjs".
    */
   serverModuleFormat: "esm" | "cjs";
+
+  /**
+   * A filename or a glob pattern to match files in the `app/routes` directory
+   * that Remix will ignore. Matching files will not be recognized as routes.
+   */
+  ignoredRouteFiles?: string;
```
Comment (kentcdodds) [inline packages/remix-dev/config.ts:91]: Could we make this an array of globs instead?

**Review by kentcdodds — COMMENTED** [review]
This looks good code-wise with one change. Needs consensus on including the feature in the first place and could use tests

**Review by kentcdodds — APPROVED** [review]
Looks good to me. Once we decide we want this I can add some docs. Cc @mjackson and @ryanflorence 
I'm in favor of this change.

**Comment by ryanflorence** [comment]
I've hesitated on this for months because I anticipated doing fancy stuff with `.css` or `.pdf`, etc. in the routes folder.

But now that we have resource routes, that seems covered.

My real question now is why have it be configuration at all? Just ignore everything except `{js,jsx,ts,tsx,md,mdx}`

**Comment by ryanflorence** [comment]
oh right, test files ...

**Comment by chaance** [comment]
Another thing I think that is nice about the config option is that if we do decide to do something fancy with certain file types in the future, it wouldn't break existing apps that have those file types ignored. They could simply opt-in by removing them from the ignore list.

**Comment by kentcdodds** [comment]
We talked about this a bit and we're thinking that the dotfiles stuff we did earlier should actually be removed (before it's released) and people can ignore dotfiles through this. The templates could be updated to ignore dotfiles via this API as well. Can you do that @chaance?

**Comment by chaance** [comment]
TBH I'm not sure how I feel about this, but I made the changes to all configs in the project to match the template defaults. Seems to add more boilerplate just to avoid a tiny bit of magic 🤷‍♂️

File: packages/remix-dev/config/routesConvention.ts [hunk 2]
```diff
@@ -21,20 +22,24 @@ export function isRouteModuleFile(filename: string): boolean {
...
 
   // First, find all route modules in app/routes
   visitFiles(path.join(appDir, "routes"), file => {
-    let routeId = createRouteId(path.join("routes", file));
-
-    if (isRouteModuleFile(file)) {
-      files[routeId] = path.join("routes", file);
+    if (
+      ignoredFilePatterns &&
+      ignoredFilePatterns.some(pattern => minimatch(file, pattern))
```
Comment (MichaelDeBoey) [inline packages/remix-dev/config/routesConvention.ts:35]: No need for a check if you give `ignoredFilePatterns` a default value of `[]`

```suggestion
      ignoredFilePatterns.some(pattern => minimatch(file, pattern))
```

File: packages/remix-dev/config/routesConvention.ts [hunk 3]
```diff
@@ -21,20 +22,24 @@ export function isRouteModuleFile(filename: string): boolean {
  * For example, a file named `app/routes/gists/$username.tsx` creates a route
  * with a path of `gists/:username`.
  */
-export function defineConventionalRoutes(appDir: string): RouteManifest {
+export function defineConventionalRoutes(
+  appDir: string,
+  ignoredFilePatterns?: string[]
```
Comment (MichaelDeBoey) [inline packages/remix-dev/config/routesConvention.ts:27]:
```suggestion
  ignoredFilePatterns?: string[] = []
```

**Review by MichaelDeBoey — CHANGES_REQUESTED** [review]

**Review by kentcdodds — APPROVED** [review]
seal of approval

**Pull request merged**